Develop rockets clickable

kieranwoodrow opened the pull request:
This feature displays the rocket info when clicked on the card. There is also testing for all the functions in the rocket info view model

kieranwoodrow committed:
added nav controller for storyboard

kieranwoodrow committed:
setup RocketInfo page in storyboard

kieranwoodrow committed:
created collection view cell file for styling

kieranwoodrow committed:
addedr ocketInfo model functions

kieranwoodrow committed:
Added structure for feature of clickable rockets

kieranwoodrow committed:
adding styling for feature

kieranwoodrow committed:
fixed constraints for heading

kieranwoodrow committed:
Update Main.storyboard

kieranwoodrow committed:
rounded photo borders

kieranwoodrow committed:
removed unnecessaary functions

kieranwoodrow committed:
Added testing for the rocketInfo feature

KirstyPallatte commented on spaceR/spaceR/Controllers/RocketInfoViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return 9
```
A suggestion going forward would be how to make this more dynamic

bgouws replied:
Once you move your array to your `ViewModel` you'll be able to make this dynamic 😄

KirstyPallatte commented on spaceR/SpacerTests/RocketTests.swift:
```diff
@@ -0,0 +1,126 @@
+    
+    class MockDelegate: ViewModelDelegate {
+        func reloadView() {
+            
```
A suggestion would be to test these delegates to see if they are called

kieranwoodrow committed:
Added tests for delegate and changed variables to private

Forgot 2 delegate tests and variables should have been private

smabhena reviewed (dismissed):
LGTM

bgouws commented on spaceR/spaceR/CustomUI/UIRocketInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,69 @@
+    func setHeadings(viewModel: RocketInfoViewModel, indexPath: Int) {
+        switch indexPath {
+        case indexPath where indexPath == 0:
+            heading = "First Flight"
```
This array of headings should live in your `ViewModel` 😄 this is coupling your code. Give me a call if you need help moving these headings to your `ViewModel`

bgouws commented on spaceR/spaceR/CustomUI/UILaunchpadTableViewCell.swift:
```diff
@@ -15,29 +15,22 @@ class UILaunchpadTableViewCell: UITableViewCell {
     }
     
-    func setLaunchpadTableViewTitle(launchpadTitle: String) {
+    func set(launchpadTitle: String) {
```
Nice naming improvement 👍

bgouws commented on spaceR/spaceR/CustomUI/UIRocketInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,69 @@
+        setHeadings(viewModel: viewModel, indexPath: indexPath)
+        set(rocketTitle: heading)
+        set(rocketInfo: info)
+        self.layer.borderColor = UIColor(red: 1, green: 0.557, blue: 0.447, alpha: 100).cgColor
```
These can be added to the style kit 😄

bgouws commented on spaceR/spaceR/CustomUI/UIRocketInfoCollectionViewCell.swift:
```diff
@@ -0,0 +1,69 @@
+        self.layer.cornerRadius = 10
+    }
+    
+    func setHeadings(viewModel: RocketInfoViewModel, indexPath: Int) {
```
This cell shouldn't really know about the `ViewModel`

bgouws commented on spaceR/spaceR/ViewModels/RocketInfoViewModel.swift:
```diff
@@ -0,0 +1,82 @@
+        return rocket?.diameter?.meters ?? 0.0
+    }
+    
+    func rocketImage() -> String {
```
This can also be a `var`

bgouws commented on spaceR/spaceR/ViewModels/RocketViewModel.swift:
```diff
@@ -25,14 +25,14 @@ class RocketViewModel {
         return allRockets?.count ?? 0
     }
     
+    func getRocket(atIndex: Int) -> Rocket? {
```
We try to avoid using the prefix "get" we ideally we want this function to read like a sentence i.e. 

`func rocket(at index: Int) -> Rocket? {`

bgouws reviewed (changes requested):
Little bit of clean up 😄

kieranwoodrow committed:
enhanced function names

KirstyPallatte reviewed (dismissed):
Looks good 👍🏻

kieranwoodrow committed:
cleaned up code

Moved headings function to view model so the function was no longer coupled by passing in the viewcontroller. Also put headings in an array so there were no string literals in the function

bgouws commented on spaceR/spaceR/ViewModels/RocketInfoViewModel.swift:
```diff
@@ -0,0 +1,127 @@
+        return URL(string: (rocket?.images.randomElement() ?? "") ?? "")
+    }
+    
+    func setHeadings(indexPath: Int) -> [String: String] {
```
This is a big improvement 😄 might still be some enhancements we can do along the way but this is where it belongs 👍

erikdvt reviewed (approved):
Looks good.
